File: Memento/c++/Memento.cpp
#include <iostream>
#include <string>
#include <stack>

class Memento
{
    public:
        virtual int GetDollars() = 0;
        virtual int GetEuro() = 0;
        virtual ~Memento() {}
};

class ExchangeMemento : public Memento
{
    private:
        int dollars;
        int euro;

    public:
        ExchangeMemento(int d, int e) : dollars(d), euro(e) {}
        int GetDollars() override;
        int GetEuro() override;
};

class Exchange
{
    private:
        int dollars;
        int euro;

    public:
        Exchange(int d, int e) : dollars(d), euro(e) {}
        void GetDollars();
        void GetEuro();
        void sell();
        void buy();
        ExchangeMemento* save();
        void restore(Memento* exchangeMemento);
};

class Memory
{
    private:
        std::stack<Memento*> history;
        Exchange* exchange;

    public:
        Memory(Exchange* ex) : exchange(ex) {}
        void backup();
        void undo();
};

int main()
{
    Exchange* exchange = new Exchange(10, 10);

    Memory* memory = new Memory(exchange);

    exchange->GetDollars();
    exchange->GetEuro();

    std::cout << "sell dollars, buy euro\n";
    exchange->sell();
    exchange->buy();

    exchange->GetDollars();
    exchange->GetEuro();

    std::cout << "save\n";
    memory->backup();

    std::cout << "sell dollars, buy euro\n";
    exchange->sell();
    exchange->buy();

    exchange->GetDollars();
    exchange->GetEuro();

    std::cout << "restoring the state\n";
    memory->undo();

    exchange->GetDollars();
    exchange->GetEuro();

    return 0;
}

int ExchangeMemento::GetDollars()
{
    return dollars;
}

int ExchangeMemento::GetEuro()
{
    return euro;
}

void Exchange::GetDollars()
{
    std::cout << "Dollars: " << this->dollars << std::endl;
}

void Exchange::GetEuro()
{
    std::cout << "Euro: " << this->euro << std::endl;
}

void Exchange::sell()
{
    this->dollars ? --this->dollars : 0;
}

void Exchange::buy()
{
    ++this->euro;
}

ExchangeMemento *Exchange::save()
{
    return new ExchangeMemento(this->dollars, this->euro);
}

void Exchange::restore(Memento *exchangeMemento)
{
    this->dollars = exchangeMemento->GetDollars();
    this->euro = exchangeMemento->GetEuro();

    delete exchangeMemento;
}

void Memory::backup()
{
    history.push(exchange->save());
}

void Memory::undo()
{
    if (history.empty()) return;

    exchange->restore(history.top());
    history.pop();
}

File: Memento/python/Memento.py
import abc
from typing import Deque


class IMemento(metaclass=abc.ABCMeta):

    @abc.abstractmethod
    def get_dollars(self) -> int:
        pass


    @abc.abstractmethod
    def get_euro(self) -> int:
        pass


class ExchangeMemento(IMemento):

    def __init__(self, d: int, e: int):
        self.__dollars = d
        self.__euro = e


    def get_dollars(self) -> int:
        return self.__dollars


    def get_euro(self) -> int:
        return  self.__euro


class Exchange:

    def __init__(self, d: int, e: int):
        self.__dollars = d
        self.__euro = e


    def get_dollars(self):
        print("Dollars: {}".format(self.__dollars))


    def get_euro(self):
        print("Euro: {}".format(self.__euro))


    def sell(self):
        if self.__dollars > 0:
            self.__dollars -= 1


    def buy(self):
        self.__euro += 1


    def save(self) -> ExchangeMemento:
        return ExchangeMemento(self.__dollars, self.__euro)


    def restore(self, exchange_memento: IMemento):
        self.__dollars = exchange_memento.get_dollars()
        self.__euro = exchange_memento.get_euro()


class Memory:

    def __init__(self, exchange: Exchange):
        self.__exchange = exchange
        self.__history: Deque[IMemento] = []


    def backup(self):
        self.__history.append(self.__exchange.save())
        

    def undo(self):
        if len(self.__history) == 0:
            return
        else:
            self.__exchange.restore(self.__history.pop())


if __name__ == "__main__":
    exchange = Exchange(d=10, e=10)

    memory = Memory(exchange)

    exchange.get_dollars()
    exchange.get_euro()

    print('sell dollars, buy euro')
    exchange.sell()
    exchange.buy()

    exchange.get_dollars()
    exchange.get_euro()

    print('Save')
    memory.backup()

    print('sell dollars, buy euro')
    exchange.sell()
    exchange.buy()

    exchange.get_dollars()
    exchange.get_euro()

    print('restoring the state')
    memory.undo()

    exchange.get_dollars()
    exchange.get_euro()